### ISSUE TITLE
Tubetool: allow to copy settings even if tube is protected

### ISSUE THREAD

**Issue by SwissalpS** [S-S-X/metatool#9]
To be tested, after list completed PR can be merged:
- [x] Teleport tube write protection works, cant be written in protected area
- [x] Teleport tube read works everywhere, protections bypassed
- [x] Teleport tube info protection works, not allowed to get list from protected area
- [x] Sorting tube write protection works, cant be written in protected area
- [x] Sorting tube read works everywhere, protections bypassed
- [x] Vacuum tube copy/paste works correctly (newly added node for tubetool)
- [x] Vacuum tube write protection works correctly (newly added node for tubetool)

If above stuff works correctly then most probably protection works correctly with all other nodes, above list covers most special and common cases where protection is either customized somehow or not customized at all (other that default config and config has unit tests).

**Comment by S-S-X** [comment]
Maybe tubes can get custom rules at some point after API, which does protection checks, is stable. 

That is probably after #7 is completed.
Currently there is open PR #8 which has not been merged yet because it causes situation where internal (within this mod) and external (any other mod) registrations would go different paths but I do have few ideas already on how to approach that problem without rewriting #8 completely.

This is something that tool should configure during registration or there could be custom callback for protection checks but currently protection check is common for any tool registrations.

Why this was disallowed at first was direct metadata manipulation which would have allowed for example to set private channels owned by other players (without very complicated checks).

**Comment by SwissalpS** [comment]
Thanks for the explanations.

**Comment by S-S-X** [comment]
Implement protection bypass settings for API #11 
Tools can then define how protections should behave.

Add protection bypass for tubetool to allow node reading after API is okayish.

**Comment by S-S-X** [comment]
To do basics:
* change item recipe so that it returns stackable metatool:toolname .. "_empty"
* change reset recipe so that it also returns metatool:toolname .. "_empty"
* after node validation at metatool:on_use
  remove 1 metatool:toolname .. "_empty" from inventory
  create new ItemStack metatool:toolname itemstack and apply data to it
  return created metatool:toolname at end of metatool:on_use

Code should go around here: https://github.com/S-S-X/metatool/blob/master/metatool/api.lua#L96-L114

**Comment by S-S-X** [comment]
After #15 is merged this can be done by simply adding `protection_bypass_read = "interact"` for mese_tube.lua here https://github.com/S-S-X/metatool/blob/master/tubetool/nodes/mese_tube.lua#L39

Protection should also now work for teleport tubes correctly and same thing should be possible for teleport_tube.lua but should be tested first.

**Comment by S-S-X** [comment]
Dependency (#11/#15) is tested and merged, it is now possible to do this.

**Comment by S-S-X** [comment]
Implemented but not tested at all, needs some testing. PR also implements vacuum tube cloning #1 and tool to use teleport tube infotext so would be good to test those too.

**Comment by S-S-X** [comment]
Teleport tube listing got implemented #33 and that caused one possible problem with protection bypass:

https://github.com/S-S-X/metatool/pull/19#issuecomment-647060898
> if this gets merged as is then anyone who can reach one protected private tp tube owned by some other player he can also get exact locations for all other tp tubes connected to that channel.
> 
> Not sure if this is acceptable or not, I think some players do not like to reveal their stuff that is marked as private.
> 
> To change this there should be separate protection bypass check before calling `info` event handler. Currently it uses same validation and priv checks that are used by `copy` / `before_read` methods.
> 
> Could simply just add `before_read_info` validation function and add similar default method with privilege bypass check / disallowed by default like `before_read` / `copy` / `before_write` / `paste`.

Above could be used to separate reading tube for cloning from reading tube for enumerating tubes in channel.

**Comment by S-S-X** [comment]
Conflicts fixed for PR, also added various other nodes to protection bypass list for reading information. All added nodes already have this information as public.

Protection bypass problem mentioned earlier should be now fixed for things that implement third `info` function.

Additional nodes with bypass read protection check privilege set to interact:
* lua controller
* lua tube
* microcontroller
* vacuum tube
* sorting tube
* injectors

For the record it is now also possible to do this through configuration by setting following key/value pairs to metatool.cfg:
```
metatool:tubetool:nodes:mese_tube:protection_bypass_read = interact
metatool:tubetool:nodes:teleport_tube:protection_bypass_read = interact
metatool:tubetool:nodes:vacuum_tube:protection_bypass_read = interact
metatool:tubetool:nodes:filter:protection_bypass_read = interact
metatool:luatool:nodes:luatube:protection_bypass_read = interact
metatool:luatool:nodes:luacontroller:protection_bypass_read = interact
metatool:luatool:nodes:microcontroller:protection_bypass_read = interact
```
